experiment/simulation/js/gate.js

```diff
@@ -0,0 +1,351 @@
+import { registerGate } from "./main.js";
+import { setPosition } from "./layout.js";
+import { halfAdder, fullAdderTest, rippleAdderTest  } from "./validator.js";
+import {jsPlumbInstance} from "./main.js";
+
+export let gates = {}; // Dictionary of gates with their IDs as keys
+window.numComponents = 0;
+//(input0,input1,carryOut,OutputOut)
+export function clearGates() {
+
+    for (let gateId in gates) {
+        delete gates[gateId];
+    }
+
+    gates = {};
+}
+
+export class Gate {
+    constructor(type) {
+        this.type = type;
+        this.id = type + "-" + window.numComponents++; // Unique ID
+        this.positionX = 0;
+        this.positionY = 0;
+        this.isConnected = false;
+        this.inputPoints = [];
+        this.outputPoints = [];
+        // connections : start/end , line_id, button_id
+        // { line_id : {button_id,start/end,}  }
+        this.inputs = []; // List of input gates
+        this.output = null; // Output value
+        this.isInput = false;
+        this.isOutput = false;
+        this.name = null;
+    }
+    // Sets the id of the gate
+    setId(id) {
+        this.id = id;
+    }
+    // Adds input to the gate
+    addInput(gate) {
+        this.inputs.push(gate);
+    }
+
+    // Removes input from the gate
+    removeInput(gate) {
+        let index = this.inputs.indexOf(gate);
+        if (index > -1) {
+            this.inputs.splice(index, 1);
+        }
+    }
+    updatePosition(id) {
+        this.positionY = window.scrollY + document.getElementById(id).getBoundingClientRect().top // Y
+
+        this.positionX = window.scrollX + document.getElementById(id).getBoundingClientRect().left // X
+    }
+    setName(name) {
+        this.name = name;
+    }
+    // generate component for the gate
+    generateComponent() {
+        let component = '';
+
+        if (this.type == "AND") {
+
+            component = '<div class="drag-drop AND" id=' + this.id + '></div>'
+
+        }
+        else if (this.type == "NOT") {
+            component = '<div class="drag-drop NOT" id=' + this.id + '></div>'
+
+        }
+        else if (this.type == "OR") {
+            component = '<div class="drag-drop OR" id=' + this.id + '></div>'
+
+        }
+        else if (this.type == "XOR") {
+            component = '<div class="drag-drop XOR" id=' + this.id + '></div>'
+
+        }
+        else if (this.type == "XNOR") {
+
+            component = '<div class="drag-drop XNOR" id=' + this.id + '></div>'
+
+        }
+        else if (this.type == "NAND") {
+
+            component = '<div class="drag-drop NAND" id=' + this.id + '></div>'
+
+        }
+        else if (this.type == "NOR") {
+
+            component = '<div class="drag-drop NOR" id=' + this.id + '></div>'

```

Comment:
probably having an if for Input and output and in the else condition
```
component = `<div class="drag-drop ${this.type}" id=' + this.id + '></div>`
```

would save some lines